compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -48,47 +52,88 @@ struct CollapseDimensionsPass
 
 /// Searches the same sequence in all the affine maps and collapses these
 /// dimensions. It only applies these to "parallel" loops without mixing them
-/// with "reduction" types.
+/// with "reduction" types. It is expected that the `genericOp` has projected
+/// permutations only as indexing maps. (Checked using `isEligibleForCollapse`).
 static SmallVector<ReassociationIndices>
 getCollapsibleLoops(linalg::GenericOp genericOp) {
   SmallVector<ReassociationIndices> contiguousLoops;
 
-  SmallVector<unsigned> pDims;
+  SmallVector<unsigned> pDims, rDims;
   genericOp.getParallelDims(pDims);
-  if (pDims.size() < 2)
-    return contiguousLoops;
-
-  llvm::SmallDenseSet<unsigned> pLoops(pDims.begin(), pDims.end());
+  genericOp.getReductionDims(rDims);
+  llvm::SmallDenseSet<unsigned> pDimsSet, rDimsSet;
+  pDimsSet.insert(pDims.begin(), pDims.end());
+  rDimsSet.insert(rDims.begin(), rDims.end());
 
   auto hasAllMapsSameSequence = [&](AffineExpr preExpr, AffineExpr nextExpr) {
+    // Check that all indexing maps of the `genericOp`
+    // - Either both `preExpr` and `nextExpr` contiguous, or
+    // - are missing in
+    // Then `preExpr` and `nextExpr` can be collapsed.
     for (AffineMap map : genericOp.getIndexingMapsArray()) {
-      bool foundSeq = false;
+      // If map has no results, no need to check.
+      if (map.getNumResults() == 0) {
+        continue;
+      }
       for (auto [index, resultExpr] : llvm::enumerate(map.getResults())) {
+        // If we find the preExpr, we should find the nextExpr.
+        if (resultExpr == preExpr) {
+          if (index == map.getNumResults() - 1) {
+            // Reached end of list. Return false;
+            return false;
+          }
+          if (map.getResult(index + 1) != nextExpr) {
+            return false;
+          }
+        }
+        // If we find nextExpr the previous one should be `prevExpr`.
+        // This is redundant check for the most part, but is cheap enough, so
+        // #YOLO
         if (resultExpr == nextExpr) {
-          foundSeq = (index > 0 && preExpr == map.getResult(index - 1));
-          break;
+          if (index == 0) {
+            // match at beginning of the list. Return false;
+            return false;
+          }
+          if (map.getResult(index - 1) != preExpr) {
+            return false;
+          }
```

Comment:
This should always be a failure right? Either index is zero or we already verified that the previous result is not preExpr in the previous iteration. I see the comment though and having this be explicit is fine.